### DISCUSSION TITLE
Multiple Guards

### DISCUSSION THREAD

**Discussion by emiliosh** [filamentphp/filament#8576] [Help] [answered]
Hi I m implementing a CMS using filament.  I have two guards setup.  One named web, and another LDAP.  I copied Login in order to try both guards at login.  And copied authenticate middleware for checking out what is the problem.  

In my tests I can use both separately but not at a time.  Because middleware always check for guards defined at filament serrvice provider.  But authguard method in panel just accepts a string, not an array.  How can I do it?

I want to try to put in a session the guard used and check if that session can be use in filament service provider.

**Comment by emiliosh** [comment, marked as answer]
I did it.  I'm using LdapRecord Library for Active Directory / LDAP auth.

Login.php
```
class Login extends \Filament\Pages\Auth\Login
{
    /**
     * @return array<int | string, string | Form>
     */
    protected function getForms(): array
    {
        return [
            'form' => $this->form(
                $this->makeForm()
                    ->schema([
                        $this->getUsernameFormComponent(),
                        $this->getPasswordFormComponent(),
                        $this->getRememberFormComponent(),
                    ])
                    ->statePath('data'),
            ),
        ];
    }

    protected function getUsernameFormComponent(): Component
    {
        return TextInput::make('username')
            ->label('Usuario')
            ->required()
            ->autocomplete()
            ->autofocus()
            ->extraInputAttributes(['tabindex' => 1]);
    }

    protected function getCredentialsFromFormData(array $data): array
    {
        return [
            'username' => $data['username'],
            'password' => $data['password'],
        ];
    }

    protected function getCredentialsFromFormDataForLdap(array $data): array
    {
        return [
            'samaccountname' => $data['username'],
            'password' => $data['password'],
        ];
    }

    public function authenticate(): ?LoginResponse
    {
        try {
            $this->rateLimit(5);
        } catch (TooManyRequestsException $exception) {
            Notification::make()
                ->title(__('filament-panels::pages/auth/login.notifications.throttled.title', [
                    'seconds' => $exception->secondsUntilAvailable,
                    'minutes' => ceil($exception->secondsUntilAvailable / 60),
                ]))
                ->body(array_key_exists('body',
                    __('filament-panels::pages/auth/login.notifications.throttled') ?: []) ? __('filament-panels::pages/auth/login.notifications.throttled.body',
                    [
                        'seconds' => $exception->secondsUntilAvailable,
                        'minutes' => ceil($exception->secondsUntilAvailable / 60),
                    ]) : null)
                ->danger()
                ->send();

            return null;
        }

        $data = $this->form->getState();

        $guard = null;
        if (!Auth::guard('ldap')
            ->attempt($this->getCredentialsFromFormDataForLdap($data),
                $data['remember'] ?? false)) {
            if (!Auth::guard('web')
                ->attempt($this->getCredentialsFromFormData($data),
                    $data['remember'] ?? false)) {
                $this->throwFailureValidationException();
            } else {
                $guard = 'web';
            }
        } else {
            $guard = 'ldap';
        }

        Auth::shouldUse($guard);
        Filament::getCurrentPanel()->authGuard($guard);
        $user = Auth::guard($guard)->user();

        if (
            ($user instanceof FilamentUser) &&
            (!$user->canAccessPanel(Filament::getCurrentPanel()))
        ) {
            Filament::auth()->logout();

            $this->throwFailureValidationException();
        }

        session()->regenerate();

        if (!is_null($guard)) {
            session(['guard' => $guard]);
        }

        return app(LoginResponse::class);
    }

    protected function throwFailureValidationException(): never
    {
        throw ValidationException::withMessages([
            'data.username' => __('filament-panels::pages/auth/login.messages.failed'),
        ]);
    }
}
```

Authenticate.php
```
class Authenticate extends \Filament\Http\Middleware\Authenticate
{
    /**
     * @param  array<string>  $guards
     */
    protected function authenticate($request, array $guards): void
    {

        Filament::getCurrentPanel()->authGuard(session('guard', 'ldap'));
        $guard = Filament::auth();

        if (!$guard->check()) {
            $this->unauthenticated($request, $guards);
            return;
        }

        $this->auth->shouldUse(session('guard', 'ldap'));

        /** @var Model $user */
        $user = $guard->user();

        $panel = Filament::getCurrentPanel();

        abort_if(
            $user instanceof FilamentUser ?
                (!$user->canAccessPanel($panel)) :
                (config('app.env') !== 'local'),
            403,
        );
    }

}
```

and finally in your AdminPanelProvider config your panel with your new Login and Authenticate Middlewares

```
...
->login(Login::class)
...
->authMiddleware([Authenticate::class]);
```

**Comment by myozin-kyaw** [comment]
Hello Brother, I've used Filament PHP notification, My Problem is provided like above Brother Emiliosh's problem. But Your solution can solve the problem but I faced with the two issues one is notification problem and second one is the permission problem I use the package Filament-Shield. I separate AdminPanelProvider and EmployeePanelProvider using the same path /admin/login page Both Panel Providers have use different Laravel Guards admin and employee. But I want to use single panel and same login page. I think  the issue is using same path() in Both Panel Provider. Could you please tell me How to solve the problem brother


    'guards' => [
        'admin' => [
            'driver' => 'session',
            'provider' => 'users',
        ],

        'employee' => [
            'driver' => 'session',
            'provider' => 'employees',
        ],
    ],

    class AdminPanelProvider extends PanelProvider
    {
        public function panel(Panel $panel): Panel
        {
            return $panel
                ->default()
                ->id('admin')
                ->path('admin')
                ->spa()
                ->colors([
                    'primary' => Color::Blue,
                ])
                ->login(HrmLogin::class)
                ->authGuard('admin')
                ->databaseNotifications()
                ->databaseTransactions()
                ->sidebarFullyCollapsibleOnDesktop()
                ->discoverResources(in: app_path('Filament/Resources'), for: 'App\\Filament\\Resources')
                ->discoverPages(in: app_path('Filament/Pages'), for: 'App\\Filament\\Pages')
                ->pages([
                    Pages\Dashboard::class,
                ])
                ->discoverWidgets(in: app_path('Filament/Widgets'), for: 'App\\Filament\\Widgets')
                ->widgets([
                    Widgets\AccountWidget::class,
                ])
                ->middleware([
                    EncryptCookies::class,
                    AddQueuedCookiesToResponse::class,
                    StartSession::class,
                    AuthenticateSession::class,
                    ShareErrorsFromSession::class,
                    VerifyCsrfToken::class,
                    SubstituteBindings::class,
                    DisableBladeIconComponents::class,
                    DispatchServingFilamentEvent::class,
                ])
                ->authMiddleware([
                    Authenticate::class,
                ])
                ->plugins([
                    FilamentShieldPlugin::make(),
                ]);
        }
    }

    class EmployeePanelProvider extends PanelProvider
    {
        public function panel(Panel $panel): Panel
        {
            return $panel
                ->default()
                ->id('admin')
                ->path('admin')
                ->spa()
                ->colors([
                    'primary' => Color::Blue,
                ])
                ->login(HrmLogin::class)
                ->authGuard('employee')
                ->databaseNotifications()
                ->databaseTransactions()
                ->sidebarFullyCollapsibleOnDesktop()
                ->discoverResources(in: app_path('Filament/Employee/Resources'), for: 'App\\Filament\\Employee\\Resources')
                ->discoverPages(in: app_path('Filament/Employee/Pages'), for: 'App\\Filament\\Employee\\Pages')
                ->pages([
                    Pages\Dashboard::class,
                ])
                ->discoverWidgets(in: app_path('Filament/Employee/Widgets'), for: 'App\\Filament\\Employee\\Widgets')
                ->widgets([
                    Widgets\AccountWidget::class,
                ])
                ->middleware([
                    EncryptCookies::class,
                    AddQueuedCookiesToResponse::class,
                    StartSession::class,
                    AuthenticateSession::class,
                    ShareErrorsFromSession::class,
                    VerifyCsrfToken::class,
                    SubstituteBindings::class,
                    DisableBladeIconComponents::class,
                    DispatchServingFilamentEvent::class,
                ])
                ->pages(Filament::getPages())
                ->resources(Filament::getResources())
                ->authMiddleware([
                    Authenticate::class,
                ])
                ->plugins([
                    FilamentShieldPlugin::make(),
                ]);
        }
    }
